horovod/common/gradient_aggregation_v2.py

```diff
@@ -0,0 +1,101 @@
+import tensorflow as tf
+
+
+class LocalGradientAggregationHelperV2:
+    def __init__(
+        self,
+        aggregation_frequency,
+        allreduce_func,
+        sparse_as_dense,
+        grad_updated_sizes_dict,
+        average_aggregated_gradients,
+    ):
+        self._allreduce_grads = allreduce_func
+
+        # How often are parameters synchronized
+        self.aggregation_frequency = aggregation_frequency
+        assert self.aggregation_frequency > 0
+
+        # Should the aggregated parameters be averaged.
+        self.average_aggregated_gradients = average_aggregated_gradients
+
+        # This is going to be N data structure holding the aggregated gradient updates
+        # for parameter updates. N is the number of parameters.
+        self.shadow_var = {}
+
+        # Used to know when to allreduce and apply gradients. We allreduce when `self.counter`
+        # is equal to `self.aggregation_frequency`. We apply gradients when `self.counter` is
+        # equal to 0.
+        self.counter = tf.Variable(initial_value=0)
+
+        self._sparse_as_dense = sparse_as_dense
+
+        # A dictionary containing the shape of each grad.
+        # This is used when gradient aggregation frequency > 1 and
+        # there are grads that have dynamically set shapes.
+        self.grad_updated_sizes_dict = grad_updated_sizes_dict
+
+        # Contains the mapping of indexes of grad updates that are
+        # not None to their index in shadow vars which only
+        # contains not None gradients. When performing gradient
+        # aggregation we have to remove them from the list of grads
+        # prior passing them into a tf.cond().
+        self.not_none_indexes = {}
+        self.num_none_grad_updates = 0
+
+    @tf.function
```

Comment:
I mean it needs it for the same reason that `compute_gradients` needs it. It's being called by the `apply_gradients` of the optimizer.

Reply:
But we don't need to add `@tf.function` since `tf.keras` will already call as a graph function with `experimental_run_tf_function` enabled.  Just ends up being a redundant call.